src/plonk/lookup/prover.rs

```diff
@@ -0,0 +1,539 @@
+        table_value: &Polynomial<C::Scalar, LagrangeCoeff>,
+    ) -> (
+        Polynomial<C::Scalar, LagrangeCoeff>,
+        Polynomial<C::Scalar, LagrangeCoeff>,
+    ) {
+        let mut input_coeffs = input_value.get_values().to_vec();
+        let table_coeffs = table_value.get_values().to_vec();
+
+        // Sort input lookup column values
+        input_coeffs.sort();
+        let permuted_input_value = Polynomial::new(input_coeffs.to_vec());
+
+        // A BTreeMap of each unique element in the table column and its count
+        let mut leftover_table_map: BTreeMap<C::Scalar, u32> =
+            table_coeffs.iter().fold(BTreeMap::new(), |mut acc, coeff| {
+                *acc.entry(*coeff).or_insert(0) += 1;
+                acc
+            });
+        let mut repeated_input_rows = vec![];
+        let mut permuted_table_coeffs = vec![C::Scalar::zero(); table_coeffs.len()];
+
+        for row in 0..input_coeffs.len() {
+            let input_value = input_coeffs[row];
+
+            // If this is the first occurence of `input_value` in the input column
+            if row == 0 || input_value != input_coeffs[row - 1] {
+                permuted_table_coeffs[row] = input_value;
+                // Remove one instance of input_value from leftover_table_map
+                if let Some(count) = leftover_table_map.get_mut(&input_value) {
+                    assert!(*count > 0);
+                    *count -= 1;
+                } else {
+                    // Panic if input_value not found
+                    panic!("Input value not found in table.");
+                }
+            // If input value is repeated
+            } else {
+                repeated_input_rows.push(row);
+            }
+        }
+
+        // Populate permuted table at unfilled rows with leftover table elements
+        for (coeff, count) in leftover_table_map.iter() {
+            for _ in 0..*count {
+                permuted_table_coeffs[repeated_input_rows.remove(0) as usize] = *coeff;
+            }
+        }
+
+        let permuted_table_value = Polynomial::new(permuted_table_coeffs.to_vec());
+
+        (permuted_input_value, permuted_table_value)
+    }
+
+    pub fn construct_product(
+        &mut self,
+        pk: &ProvingKey<C>,
+        params: &Params<C>,
+        beta: C::Scalar,
+        gamma: C::Scalar,
+        theta: C::Scalar,
+        advice_values: &[Polynomial<C::Scalar, LagrangeCoeff>],
+        fixed_values: &[Polynomial<C::Scalar, LagrangeCoeff>],
+    ) -> Product<C> {
+        let permuted = self.permuted.clone().unwrap();
+        let unpermuted_input_values: Vec<Polynomial<C::Scalar, LagrangeCoeff>> = self
+            .lookup
+            .input_columns
+            .iter()
+            .map(|&input| match input.column_type {
+                Any::Advice => advice_values[input.index].clone(),
+                Any::Fixed => fixed_values[input.index].clone(),
+                _ => unreachable!(),
```

Comment:
I don't think this is unreachable, since `ConstraintSystem::lookup` allows `Column<Any>` for the input and table columns. I can't immediately think of a use-case for an aux column in a lookup, but I also can't think of a reason to prohibit it.